controllers/ctltools.js

```diff
@@ -2,301 +2,225 @@
+        case 'basic':
+            break;
+        default:
+            throw new Error('Unknown tool config type');
+    }
 
-const sanitizeTool = (req, create=false) => {
-  req.body.key = req.body.key || null;
-  req.body.resource_link_id = req.body.resource_link_id || null;
-  req.body.toolType = req.body.toolType || null;
-  req.body.lti_version = req.body.lti_version || null;
-  req.body.lti_message_type = req.body.lti_message_type || null;
-  req.body.secret = (req.body.secret === PASSWORD ? undefined : req.body.secret);
-  req.body.isTemplate = true;
-  if(create || !req.body.isLocal) { // non-local (LTI) tools can be updated forever, local (OAuth2) only during creation
-    req.body.oAuthClientId = req.body.oAuthClientId || "";
-  } else {
-    req.body.oAuthClientId = undefined; // undefine the property prohibits database update
-  }
-  req.body.isLocal = (create ? !!req.body.isLocal : undefined);
-  req.body.scope = req.body.scope || "openid offline";
-  req.body.skipConsent = !!req.body.skipConsent;
-  req.body.openNewTab = !!req.body.openNewTab;
-  req.body.frontchannel_logout_uri = req.body.frontchannel_logout_uri || null;
-  req.body.isHidden = !!req.body.isHidden;
-  return req;
+    if(body.parameters && Array.isArray(body.parameters)) {
+        body.parameters.forEach((param) => {
+            param.default = param.default || undefined;
+            param.regex = param.regex || undefined;
+        });
+    }
+
+    trimWhitespaces(body);
+
+    return body;
 };
 
-const createTool = (req, service, next) => {
-  api(req).post('/' + service + '/', {
-    json: req.body
-  }).then(tool => {
-    next();
-  }).catch(err => {
-    if(req.body.isLocal) {
-      api(req, { version: HYDRA_VERSION }).delete(`/oauth2/clients/${req.body.oAuthClientId}`).then(_ => {
+const getUpdateHandler = (req, res, next) => {
+    req.body = sanitizeToolInputs(req.body);
+
+    api(req, { version: 'v3' }).put(`/tools/${req.params.id}`, {
+        json: req.body
+    }).then(() => {
+        res.redirect(req.header('Referer'));
+    }).catch(err => {
         next(err);
-      });
-    }
-    next(err);
-  });
+    });
 };
 
-const getCreateHandler = (service) => {
-    return function (req, res, next) {
-      req = sanitizeTool(req, true);
-      if(req.body.isLocal) {
-        return api(req, { version: HYDRA_VERSION }).post('/oauth2/clients/', {
-          json: getClient(req.body, true)
-        }).then(response => {
-          req.body.oAuthClientId = response.client_id;
-          createTool(req, service, next);
-        }).catch(err => {
-          next(err);
-        });
-      } else {
-        createTool(req, service, next);
-      }
-    };
+const getDetailHandler = (req, res, next) => {
+    api(req, { version: 'v3' }).get(`/tools/${req.params.id}`).then(data => {
+        if (data.config.type === 'oauth2') {
+            data.config.redirectUris = data.config.redirectUris.join(';');
+            data.config.clientSecret = PASSWORD;
```

Comment:
I assume this is the prefill the user sees in the password field? is there no better way to do this on the field itself? I cant believe its a best practise to manually set this on pageload, and then sanitize it on submit...

Reply:
it was the previous practice to do it like that. This is just copied code for now. Once we start to properly implement the new functionalities we will take a look into improving this code